docs/src/lib/indexing.md

```diff
@@ -21,66 +21,111 @@ The rules for a valid type of index into a column are the following:
+    * a regular expression, which gets expanded to a vector of matching column names;
+    * a `Not` expression;
+    * a colon `:`.
 
 The rules for a valid type of index into a row are the following:
 * a value, later denoted as `row`:
     * an `Integer` that is not `Bool`;
 * a vector, later denoted as `rows`:
     * a vector of `Integer` other than `Bool` (does not have to be a subtype of `AbstractVector{<:Integer}`);
     * a vector of `Bool` that has to be a subtype of `AbstractVector{Bool}`;
-    * a colon.
+    * a `Not` expression;
+* a colon `:`;
+* an exclamation mark `!`.
+
+Additionally it is allowed to index into an `AbstractDataFrame` using a two dimensional `CartesianIndex`.
 
 In the descriptions below `df` represents a `DataFrame`, `sdf` is a `SubDataFrame` and `dfr` is a `DataFrameRow`.
 
-## `getindex`
+## `getindex` and `view`
 
 The following list specifies return types of `getindex` operations depending on argument types.
 
-In all operations copying vectors is avoided where possible.
-If it is performed a description explicitly mentions that the data is *copied*.
+In particular a description explicitly mentions that the data is *copied* or *reused without copying*.
 
 For performance reasons, accessing, via `getindex` or `view`, a single `row` and multiple `cols` of a `DataFrame`, a `SubDataFrame` or a `DataFrameRow` always returns a `DataFrameRow` (which is a view-like type).
 
 `DataFrame`:
-* `df[col]` -> the vector contained in column `col`;
-* `df[cols]` -> a freshly allocated `DataFrame` containing the copies of vectors contained in columns `cols`;
-* `df[row, col]` -> the value contained in row `row` of column `col`, the same as `df[col][row]`;
 * `df[CartesianIndex(row, col)]` -> the same as `df[row,col]`;
-* `df[row, cols]` -> a `DataFrameRow` with parent `df` if `cols` is a colon and `df[cols]` otherwise;
-* `df[rows, col]` -> a copy of the vector `df[col]` with only the entries corresponding to `rows` selected, the same as `df[col][rows]`;
-* `df[rows, cols]` -> a `DataFrame` containing copies of columns `cols` with only the entries corresponding to `rows` selected.
-* `@view df[col]` -> the vector contained in column `col` (this is equivalent to `df[col]`);
-* `@view df[cols]` -> a `SubDataFrame` with parent `df` if `cols` is a colon and `df[cols]` otherwise;
-* `@view df[row, col]` -> a `0`-dimensional view into `df[col]`, the same as `view(df[col], row)`;
-* `@view df[row, cols]` -> a `DataFrameRow` with parent `df` if `cols` is a colon and `df[cols]` otherwise;
-* `@view df[rows, col]` -> a view into `df[col]` with `rows` selected, the same as `view(df[col], rows)`;
-* `@view df[rows, cols]` -> a `SubDataFrame` with `rows` selected with parent `df` if `cols` is a colon and `df[cols]` otherwise.
+* `df[row, col]` -> the value contained in row `row` of column `col`, the same as `df[!, col][row]`;
+* `df[row, cols]` -> a `DataFrameRow` with parent `df` if `cols` is a colon and parent `df[!, cols]` otherwise;
```

Comment:
This is a delicate point. `df[!, :]` returns a copy of `df` and this is different because:
* it would be more expensive computationally (and we want to create `dfr` very cheaply)
* it would mean that when we mutate `df` (add/remove/rename columns) then `dfr` would not be mutated and it is mutated

Reply:
Ah right `df[!, :]` doesn't copy vectors but it copies the data frame.